Active Record Query Interface > 2. Conditions (#371)
campuscode/rails-guides-pt-BR: pt-BR/active_record_querying.md

O método `find_in_batches` aceita as mesmas opção que o `find_each`

Condições
----------

> O método `where`  permite que você especifique condições para limitar os registro retornados, representando o `where`-part da instrução SQL. Condições podem ser especificado como uma string, array(lista), ou hash.

Author: Corrigido

### Condições de Strings Puras

> Se você gostaria de adicionar condições para sua procura, você poderia apenas especifica-los lá, somente como `Client.where("orders_count = '2'")`. Isso encontrará todos os clientes onde o campo `ordes_count` que tenha o valor igual a 2.

Author: Corrigido

> WARNING: Construindo sua própria condição como strings pura pode te deixar vulnerável a façanha de injeção SQL. Por exemplo, `Client.where("first_name LIKE '%#{params[:first_name]}%'")` não é seguro. Veja a pŕoxima seção para a maneira de lidar com  condições usando array.

Reviewer: Achei que "façanhas" como tradução para "exploit" ficou um pouco fora do contexto. Fiz uma sugestão...

```suggestion
WARNING: Construindo sua própria condição como *string* pura pode te deixar vulnerável a ataques de injeção SQL. Por exemplo, `Client.where("first_name LIKE '%#{params[:first_name]}%'")` não é seguro. Veja a próxima seção para a maneira de lidar com condições usando *array*.
```

Author: Corrigido

### Condições de Array

> Agora, se esse número pudesse variar, digamos como um argumento de algum lugar? O achado então levaria a forma:

Reviewer: ```suggestion
Agora, se esse número pudesse variar, digamos como um argumento de algum lugar? O comando da busca então levaria a forma:
```

Author: Corrigido

```ruby
Client.where("orders_count = ?", params[:orders])
```

> Active Record tomará o primeiro argumento como a string de condições e quaisquer argumento adicional vai substituir os pontos de interrogação `(?)` nele.

Reviewer: ```suggestion
*Active Record* tomará o primeiro argumento como a *string* de condições e quaisquer argumentos adicionais vão substituir os pontos de interrogação `(?)` nele.
```

Author: Corrigido

> Se você quer especificar muiltiplas condições:

Reviewer: ```suggestion
Se você quer especificar múltiplas condições:
```

Author: Corrigido

```ruby
Client.where("orders_count = ? AND locked = ?", params[:orders], false)
```

Neste exemplo, o primeiro ponto de interrogação será substituído com o valor em `params[:orders]` e o segundo será substituído com a representação SQL para `false`, que depende do adaptador.

Este código é altamente preferível:

```ruby
Client.where("orders_count = ?", params[:orders])
```

Para este código:

```ruby
Client.where("orders_count = #{params[:orders]}")
```

> Por causa da segurança do argumento. Colocando a variável dentro da condição de string parará a variable para o banco de dados **como se encontra**. Isto significa que isso será uma variável sem escape diretament de um user que pode ter intenções maliciosas. Se você fizer isso, you coloca todo seu banco de dados em risco porque uma vez que um user descobre, ele pode explorar seu banco de dados e pode fazer qualquer coisa com ele. Nunca, jamais, coloque seus argumentos diretamenta dentro da condição de string.

Reviewer: ```suggestion
devido à segurança do argumento. Colocando a variável dentro da condição de *string*, passará a variável para o banco de dados **como se encontra**. Isto significa que será uma variável sem escape diretamente de um usuário que pode ter intenções maliciosas. Se você fizer isso, coloca todo seu banco de dados em risco, porque uma vez que um usuário descobre que pode explorar seu banco de dados, ele pode fazer qualquer coisa com ele. Nunca, jamais, coloque seus argumentos diretamente dentro da condição de *string*.
```

Author: Corrigido

> TIP: Para mais informações sobre os perigos  da injeção de SQL, veja em [Ruby on Rails Security Guide](https://guides.rubyonrails.org/security.html#sql-injection) / [Ruby on Rails Security Guide PT-Br](security.html#sql-injection)

Reviewer: ```suggestion
TIP: Para mais informações sobre os perigos da injeção de SQL, veja em [Ruby on Rails Security Guide](https://guides.rubyonrails.org/security.html#sql-injection) / [Ruby on Rails Security Guide PT-Br](security.html#sql-injection)
```

Author: Corrigido

> #### Placeholder na condição

Reviewer 2: ```suggestion
#### Condições com *Placeholder*
```

Aqui fiquei um pouco na dúvida mas seria legal mantermos o mesmo estilo do título anterior, acha que faz sentido?

Author: Siim, ficou bem melhor kkk

> Similar ao estilo de substituição `(?)` dos parâmetros, você também pode especificar chaves em sua condição de string junto com um hash de chaves/valores(keys/values) correspondentes:

Reviewer: ```suggestion
Similar ao estilo de substituição `(?)` dos parâmetros, você também pode especificar chaves em sua condição de *string* junto com uma *hash* de chaves/valores (*keys/values*) correspondentes:
```

Author: Corrigido

```ruby
Client.where("created_at >= :start_date AND created_at <= :end_date",
  {start_date: params[:start_date], end_date: params[:end_date]})
```

> Isso torna a legibilidade mais clara se você tem um grande número de condições variaveis.

Reviewer 2: ```suggestion
Isso torna a legibilidade mais clara se você tem um grande número de condições variáveis.
```

Aqui acho que tem um typo, pode dar uma olhada?

### Condições de Hash

> Active Record também permite que você passe em condições de hash o que pode aumentar a legibilidade de suas sintaxes de condições. Com condições de hash, você passa em uma hash com chaves(keys) dos campos que deseja qualificados e os valores(values) de como deseja qualificar-los:

Reviewer: ```suggestion
*Active Record* também permite que você passe em condições de *hash* o que pode aumentar a legibilidade de suas sintaxes de condições. Com condições de *hash*, você passa em uma *hash* com chaves (*keys*) dos campos que deseja qualificados e os valores (*values*) de como deseja qualificá-los:
```

Author: Corrigidos

> NOTE: Apenas igualdade, intervalo, e subconjunto são possiveis com as condições de hash.

Reviewer: ```suggestion
NOTE: Apenas igualdade, intervalo, e subconjunto são possiveis com as condições de *hash*.
```

Author: Corrigido

> NOTE: Apenas igualdade, intervalo, e subconjunto são possiveis com as condições de *hash*.

Reviewer 2: ```suggestion
NOTE: Apenas igualdade, intervalo, e subconjunto são possíveis com as condições de *hash*.
```

Aqui acho que ficou um typo, pode dar uma olhada?

#### Condições de igualdade

```ruby
Client.where(locked: true)
```

Isso irá gerar um SQL como este:

```sql
SELECT * FROM clients WHERE (clients.locked = 1)
```

> O nome do campo também pode ser uma string:

Reviewer: ```suggestion
O nome do campo também pode ser uma *string*:
```

Author: Corrigido

```ruby
Client.where('locked' => true)
```

> No caso de um relacionamento `belongs_to`, uma chave de associação pode ser usada para especificar o model se um objeto Active Record for usado como o valor. Este metodo também funciona com relacionamentos polimórficos.

Reviewer 2: ```suggestion
No caso de um relacionamento `belongs_to`, uma chave de associação pode ser usada para especificar o model se um objeto Active Record for usado como o valor. Este método também funciona com relacionamentos polimórficos.
```

Aqui acho que ficou um typo, pode dar uma olhada?

```ruby
Article.where(author: author)
Author.joins(:articles).where(articles: { author: author })
```

#### Condições de intervalos

```ruby
Client.where(created_at: (Time.now.midnight - 1.day)..Time.now.midnight)
```

> Isso irá encontrar todos clientes criado ontem usando uma instrução SQL `BETWEEN`:

Reviewer: ```suggestion
Isso irá encontrar todos clientes criados ontem usando uma instrução SQL `BETWEEN`:
```

Author: Corrigido

```sql
SELECT * FROM clients WHERE (clients.created_at BETWEEN '2008-12-21 00:00:00' AND '2008-12-22 00:00:00')
```

> Essa demonstração uma sintaxe mais curta para examplos em [Condições de Array](#condiçoes-de-array)

Reviewer: ```suggestion
Isso demonstra uma sintaxe mais curta para exemplos em [Condições de Array](#condiçoes-de-array)
```

Author: Corrigido

> Isso demonstra uma sintaxe mais curta para exemplos em [Condições de Array](#condiçoes-de-array)

Reviewer 2: ```suggestion
Isso demonstra uma sintaxe mais curta para exemplos em [Condições de Array](#condicoes-de-array)
```

Os links perdem os caracteres especiais então deixei a sugestão de como vai ficar 😄

Author: Vdd, deixei sem o `~` no `O`, mas esqueci do `ç`, vlw

#### Subconjunto de Condições

> Se você deseja procurar registros usando a expressão `IN` pode passar uma array para o hash de condições:

Reviewer: ```suggestion
Se você deseja procurar registros usando a expressão `IN` pode passar um *array* para a *hash* de condições:
```

Author: Corrigido

```ruby
Client.where(orders_count: [1,3,5])
```

> Isso código irá gerar um SQL como este:

Reviewer: ```suggestion
Esse código irá gerar um SQL como este:
```

Author: Corrigido

```sql
SELECT * FROM clients WHERE (clients.orders_count IN (1,3,5))
```

### Condições NOT

> Consultas SQL `NOT` podem ser construidas por `where.not`:

Reviewer 2: ```suggestion
Consultas SQL `NOT` podem ser construídas por `where.not`:
```

Aqui acho que ficou um typo, pode dar uma olhada?

```ruby
Client.where.not(locked: true)
```

Em outras palavras, essa consulta pode ser gerada chamando `where` sem nenhum argumento, então imediatamente encadeie com condições `not` passando `where`. Isso irá gerar SQL como este:

```sql
SELECT * FROM clients WHERE (clients.locked != 1)
```

### Condições OR

> Condições `OR` entre duas relações podem ser construidas chamando `or` na primeira relação, e passando o segundo como um argumento.

Reviewer 2: ```suggestion
Condições `OR` entre duas relações podem ser construídas chamando `or` na primeira relação, e passando o segundo como um argumento.
```

Aqui acho que ficou um typo, pode dar uma olhada?
